src/main/scala/com/wavesplatform/UtxPool.scala

```diff
@@ -0,0 +1,90 @@
+package com.wavesplatform
+
+import java.util.concurrent.ConcurrentHashMap
+
+import cats.Monoid
+import com.wavesplatform.network._
+import com.wavesplatform.settings.{FunctionalitySettings, UtxSettings}
+import com.wavesplatform.state2.diffs.TransactionDiffer
+import com.wavesplatform.state2.reader.{CompositeStateReader, StateReader}
+import com.wavesplatform.state2.{ByteStr, Diff}
+import io.netty.channel.Channel
+import io.netty.channel.group.ChannelGroup
+import scorex.consensus.TransactionsOrdering
+import scorex.transaction.ValidationError.GenericError
+import scorex.transaction.{FeeCalculator, Transaction, ValidationError}
+import scorex.utils.{ScorexLogging, Time}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.duration._
+import scala.util.{Left, Right}
+
+
+class UtxPool(
+    allChannels: ChannelGroup,
+    time: Time,
+    stateReader: StateReader,
+    feeCalculator: FeeCalculator,
+    fs: FunctionalitySettings,
+    utxSettings: UtxSettings) extends ScorexLogging {
+
+  private val transactions = new ConcurrentHashMap[ByteStr, Transaction](utxSettings.maxSize / 2)
+
+  private def validate(t: Transaction): Either[ValidationError, Transaction] =
+    TransactionDiffer.apply(fs, time.correctedTime(), stateReader.height)(stateReader, t)
+      .map(_ => t)
+
+  private def collectValidTransactions(currentTs: Long): Seq[Transaction] = {
+    val differ = TransactionDiffer.apply(fs, currentTs, stateReader.height) _
+
+    val (invalidTxs, validTxs, _) = transactions.asScala
+      .values.toSeq
+      .sorted(TransactionsOrdering.InUTXPool)
+      .foldLeft((Seq.newBuilder[ByteStr], ArrayBuffer.empty[Transaction], Monoid[Diff].empty)) {
```

Comment:
`Seq.newBuilder` does exactly that: prepends new head to a cached list, and then reverses it when `result()` is called. You just don't need to do it yourself. As for `ArrayByffer` of transactions, I need to check its length, so using `Seq` builder here and reverse and then traverse it just to count elements seems wasteful.